### PR TITLE
Pipenv

### PR DESCRIPTION
**Pull request by thatandromeda**
## Status
**READY FOR REVIEW**

#### What does this PR do?
* Replaces `requirements.txt` and its associated workflows with `pipenv`.
* Upgrades mock and pytest.
* Updates `tox.ini` for compatibility with `pipenv`.
* Discards the mostly duplicative `tox.mysql.ini` in favor of adding a passenv statement to tox.ini; this ensures that tests run in the database used by the test environment (which means mysql when run through docker).
* Ensures that the new `tox.ini` works with both localhost and docker environments.

Why Pipenv? It is just better:
* it results in more deterministic builds, by locking versions of all the dependencies and their dependencies, and hashing them for extra peace of mind;
* it also replaces `virtualenv`, and has a much easier workflow
* it's the emerging standard, and officially recommended by PyPI
* it can still generate `requirements.txt` files if those are required for integration with existing build processes

[Read the docs](https://pipenv.readthedocs.io/) for more details.

#### What are the relevant tickets?
- [x] https://github.com/unt-libraries/django-premis-event-service/projects/1#card-9791562
- [x] https://github.com/unt-libraries/django-premis-event-service/projects/1#card-9786216
- [x] https://github.com/unt-libraries/django-premis-event-service/projects/1#card-9819326
- [x] https://github.com/unt-libraries/django-premis-event-service/projects/1#card-9794269
- [x] https://github.com/unt-libraries/django-premis-event-service/projects/1#card-9820365

#### Requires Database Migrations?
NO

#### Includes new or updated dependencies?
YES

### CODE REVIEW

**Comment by thatandromeda** [comment]
@ldko This is now ready for further review. (Do you prefer your merges squashed or unsquashed, when we get to that point?)

**Comment by ldko** [comment]
@thatandromeda as a group here we tend not to squash our merges. Sometimes it is nice to trace back to things that were tried and reviewed and changed in certain places. There are cases when some of us do squash commits depending on what is in the set of commits. Perhaps if no one has asked for changes or the separate commits are trivial. In the case of this PR, I think squashing or not is fine. Hopefully that answers your question, if not very clearly :)...

**Comment by thatandromeda** [comment]
Totally makes sense. I'll leave it unsquashed after review.

File: README.rst [hunk 1]
```diff
@@ -81,26 +83,25 @@ Clone the repository
...
+    (premis-event-service) $ exit (to leave the virtualenv)
 
-.. code-block :: sh
+If your Python 2.7 has a different minor version number than 15, you can
+specify that and it should still work. However, if you're using OS X Sierra or
+later, you will need to specify 2.7.15, because a change in Mac handling of
+OpenSSL broke pip for earlier versions of Python.
 
-    (premis-event-service) $ pip install -r requirements.txt # install dependencies from text file
+If you need to generate a requirements.txt file, you can do so with ``pipenv lock -r``.
```
Comment (ldko) [inline README.rst]: could specify `pipenv lock -r > requirements.txt`

File: Pipfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+Django = "~=1.8"
+
+[dev-packages]
+django-debug-toolbar = "*"
+tox = "*"
+pytest = "==3.5.1"
+pytest-django = "==2.9.1"
+mock = "==1.3.0"
+factory-boy = "==2.5.2"
+"flake8" = "*"
```
Comment (ldko) [inline Pipfile]: Does this mean something different having `"flake8"` in double quotes here?

**Comment by somexpert** [comment]
I'm still quite new to pipenv and docker in general, but I was able to get the docker container working (after some trial and error) by adding the Pipfile and Pipfile.lock files to the image. Not sure if it could be done a better way, but I just added `ADD Pipfile /app/` and `ADD Pipfile.lock /app/` to the Dockerfile before the `pipenv` command was run.

File: Pipfile [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+url = "https://pypi.org/simple"
+verify_ssl = true
+name = "pypi"
+
+[packages]
+lxml = "~=3.3.0"
+codalib = "==1.0.3"
+sqlparse = "*"
+mysql-python = "*"
+Django = "~=1.8"
```
[the remark this answers is not in the document]
Comment (thatandromeda) [reply Pipfile]: The intention is 1.8 - the lockfile ended up out of sync while I was fighting with Docker, and I'll fix that.

File: README.rst [hunk 4]
```diff
@@ -81,26 +83,25 @@ Clone the repository
...
 
+.. _pipenv: https://pipenv.readthedocs.io/en/latest/
 
-Install the requirements using pip_
-"""""""""""""""""""""""""""""""""""
+.. code-block :: sh
 
-.. _pip: https://pip.pypa.io/en/stable/
+    (premis-event-service) $ pipenv --python 2.7.15 (to create the virtualenv)
+    (premis-event-service) $ pipenv install --dev
```
[the remark this answers is not in the document]
Comment (thatandromeda) [reply README.rst]: 👍

File: README.rst [hunk 5]
```diff
@@ -81,26 +83,25 @@ Clone the repository
...
+    (premis-event-service) $ pipenv shell (to enter the virtualenv)
+    (premis-event-service) $ exit (to leave the virtualenv)
```
[the remark this answers is not in the document]
Comment (thatandromeda) [reply README.rst]: 👍

File: tox.ini [hunk 6]
```diff
@@ -15,21 +12,53 @@ envlist=
...
-    django110: Django~=1.10.0
-    djangomaster: https://github.com/django/django/archive/master.tar.gz
-commands = py.test
+    pipenv
+commands =
+    pipenv install --dev --ignore-pipfile
+
+    # Override pipenv version of Django as needed by test envs. Use pip so that
+    # Pipfile.lock doesn't end up changed by this process.
+    django17: pip install Django~=1.7.0
```
[the remark this answers is not in the document]
Comment (thatandromeda) [reply tox.ini]: Somewhat annoyingly, pipenv only supports `[packages]` and `[dev-packages]`, not a testing-only environment. I'd noticed the same thing. I'll drop 1.7 support.

File: Dockerfile [hunk 7]
```diff
@@ -12,9 +12,8 @@ WORKDIR /app
 
 RUN apt-get update -qq && apt-get install -y mysql-client netcat
 
-ADD requirements/ /app/requirements
-ADD requirements.txt /app/
-RUN pip install -r requirements.txt
+RUN pip install pipenv
```
[the remark this answers is not in the document]
Comment (thatandromeda) [reply Dockerfile]: 👍 I didn't encounter that error so it's helpful to know you do. I've added your command to the Dockerfile.

File: Pipfile (hunk 2 above)
Comment (thatandromeda) [reply Pipfile]: `pipenv install` updates the Pipfile automatically and this is what it does for this package. I don't know why.

**Comment by thatandromeda** [comment]
I'm pretty new to Docker myself and I'm finding the amount of trial and error is enough to make me nervous about repeatability - I'm really glad you both are testing it on your systems as well. But with my latest commits I've responded to your PR comments; found a bug in my Pipfile syntax which I think explains some things; and got tests (both tox and docker+tox) passing on localhost. Tell me if this works for you, and I'll see how CI feels about it.

**Comment by ldko** [comment]
This is working for me now both with a local dev environment and Docker! I noticed a couple things in README.rst that should be updated to go along with the changes in this PR. References to Django 1.7 should be changed to 1.8, and the part about "However, if the requirements files change, it is important that you rebuild the app container" should be made relevant to using Pipenv/Pipfile.

**Comment by somexpert** [comment]
It's all working for me now too :smile:

**Comment by thatandromeda** [comment]
@ldko ready for merge!

**Comment by ldko** [comment]
Thanks @thatandromeda !

**Pull request merged**